File: ruby/rails7-sidekiq-opentelemetry/app/config/initializers/opentelemetry.rb
require "opentelemetry/sdk"
require "opentelemetry/instrumentation/all"

OpenTelemetry::SDK.configure do |c|
  c.service_name =
    if Sidekiq.server?
      "Sidekiq worker"
    else
      "Rails server"
    end
  c.resource = OpenTelemetry::SDK::Resources::Resource.create(
    "appsignal.config.app_name" => ENV.fetch("APPSIGNAL_APP_NAME"),
    "appsignal.config.app_environment" => ENV.fetch("APPSIGNAL_APP_ENV"),
    "appsignal.config.push_api_key" => ENV.fetch("APPSIGNAL_PUSH_API_KEY"),
    "appsignal.config.app_path" => Rails.root.to_s
  )
  c.use_all(
    "OpenTelemetry::Instrumentation::Sidekiq" => {
      :span_naming => :job_class
    },
  ) # enables all instrumentation
end

puts "!!! opentelemetry start"

File: javascript/react-otel/app/src/opentelemetry.js
import { WebTracerProvider } from "@opentelemetry/sdk-trace-web";
import { Resource } from "@opentelemetry/resources";
import { SemanticResourceAttributes } from "@opentelemetry/semantic-conventions";
import { ZoneContextManager } from "@opentelemetry/context-zone";
import { BatchSpanProcessor, SimpleSpanProcessor, ConsoleSpanExporter } from "@opentelemetry/sdk-trace-base";
import { OTLPTraceExporter } from "@opentelemetry/exporter-trace-otlp-proto";
import { registerInstrumentations } from "@opentelemetry/instrumentation";
import { XMLHttpRequestInstrumentation } from "@opentelemetry/instrumentation-xml-http-request";
import { FetchInstrumentation } from "@opentelemetry/instrumentation-fetch";
import { UserInteractionInstrumentation } from "@opentelemetry/instrumentation-user-interaction";
import { DocumentLoadInstrumentation } from "@opentelemetry/instrumentation-document-load";
import { LongTaskInstrumentation } from "@opentelemetry/instrumentation-long-task";
import * as api from "@opentelemetry/api";

export const initializeTracing = () => {
  api.diag.setLogger(
    new api.DiagConsoleLogger(),
    api.DiagLogLevel.ALL,
  );

  const provider = new WebTracerProvider({
    resource: new Resource({
      "appsignal.config.revision": "24-11-05",
      "appsignal.config.app_name": "javascript/react-otel",
      "appsignal.config.app_environment": "development",
      "appsignal.config.push_api_key": "",
      "appsignal.config.language_integration": "javascript",
      [SemanticResourceAttributes.SERVICE_NAME]: "javascript-react-otel",
    }),
  });

  // Register all common web instrumentations at once
  registerInstrumentations({
    instrumentations: [
      new XMLHttpRequestInstrumentation({
        propagateTraceHeaderCorsUrls: /.*/,
        debugWrappers: true,
      }),
      new FetchInstrumentation({
        propagateTraceHeaderCorsUrls: /.*/,
        debugWrappers: true,
      }),
      new UserInteractionInstrumentation(),
      new DocumentLoadInstrumentation(),
      new LongTaskInstrumentation(),
    ],
  });

  const collector = new OTLPTraceExporter({
    url: "http://localhost:3030/http://appsignal-agent:8099/enriched/v1/traces",
  });

  const consoleExporter = new ConsoleSpanExporter();

  provider.addSpanProcessor(new BatchSpanProcessor(collector));
  provider.addSpanProcessor(new SimpleSpanProcessor(consoleExporter));
  provider.register({ contextManager: new ZoneContextManager() });

  return provider.getTracer("react-tracer");
};
